### DISCUSSION TITLE
Better audio clip management for looping and comping

### DISCUSSION THREAD

**Discussion by seangoodvibes** [SynthstromAudible/DelugeFirmware#504] [Ideas]
Currently when you create an audio clip and overdub it creates multiple audio clips that fill up your song view pretty quickly.

Suggested improvement:
- group all related audio clips under a single audio clip. The different "takes" of that audio clip would be loaded into the sidebar of that audio clip, in the exact way sounds are loaded into rows in a kit instrument clip.

This would enable you to then use the sidebar in the audio clip to quickly navigate between all related audio clip takes and adjust the loop points

This would then set us up to create comping functionality that could look at all the takes and apply a cross fade between them based on the loop points to create one audio clip that is comprised of the best parts of multiple takes.

**Comment by 0beron** [comment]
Would this replace the existing method completely, or be requested somehow / toggled as a community feature? The existing workflow of filling up additional clips supports live looping, wherein each clip is different as opposed to repeated takes of the same material. Having them in separate clips leaves them accessible to control with the launch/status pads, and tidying them automatically into a single audio clip would prevent this.

**Reply by seangoodvibes** [reply]
I think some thought would need to be taken for sure as to how this change is implemented.

I know there is talk of a live looping grid mode for song view that would also be factored into this discussion.

@trappar and @m-m-adams can provide more insight onto the benefits of this idea. Maybe they can also weigh in on your viewpoint about wanting to keep the takes "exposed" as opposed to hidden in a single clip.

**Reply by m-m-adams** [reply]
I was actually thinking of this mainly as a benefit to live looping - the current workflow doesn't permit you to do things like reverse playback or double playback speed of all layers of a loop simultaneously, which is a feature I sorely miss from other loopers. Having the audio tracks within a container clip would let you do whole loop effects much more easily. It'll also resolve the random issues that can arise by accidentally changing which clip is the 'active' one for the next loop by dramatically reducing the number of visible clips. It'd be something similar to a kit, where there is an individual mute/launch pad for each row, but it would hold full loop length audio clips instead of samples

**Reply by m-m-adams** [reply]
The comping idea was really a bonus - if you have a toggle to make all audio clips within a loop-clip mutually exclusive then there's an easy way to do comping by just selecting the row to be played at any given time in the loop-clip

**Comment by m-m-adams** [comment]
I guess to summarize what the proposed workflow would be:

1. Loop command
    - Create a new clip below the current one on a new instrument
2. Overdub command
    - Convert the audio/midi/kit clip into a layer container
    - The layer container would have each loop/layer in a row, like a kit that contains audio loops or sequences
    - The layer container could have a loop mode where all layers play at once, and an exclusive switch where one plays at a time
    - In the exclusive mode you could highlight pads you'd like to use from each layer to allow comping

**Reply by Eighthours2023** [reply]
This sounds great!

**Reply by kozmikyak** [reply]
I second all of this.  I'm seriously panning to get a Deluge after finally "getting" its features due to viewing tutorials and such, and want to use it for live looping, and will be able to, but it would be a benefit to be able to overdub onto a single "container".  I wouldn't want that "container" to be forced to be a section, I'd want it to be like a track in that you can mute it or play it as _part_ of a section.  The use case is building up chords with 7-string electric cello over the course of a few overdubs, then treating that "chords" track like a single track.  Or "clip" if that's the proper terminology for it.  Sorry if I'm intruding.

**Reply by kozmikyak** [reply]
Another option for this would be to allow for a resample of a set of overdubs, and replacement of the individually overdubbed clips with that resample.  But that would be potentially "destructive" and I do like the ability to have the original overdub tracks available.

**Comment by sichtbeton** [comment]
splitting a sample / audio clip would be very interesting too regarding comping.

**Reply by sichtbeton** [reply]
also: could a kit just hold audio clips? this way you have your layers already.

**Comment by duncanrmi** [comment]
I made the mistake over the summer of setting the deluge in audio loop record/overdub mode so that it kept creating new clips & filling up the grid... I wanted to see how far it would go before it crashed, so I created a ten second loop & then once it was recording, I turned the tempo way up.... somewhere in the thousands. the damn thing just kept going until there were hundreds of clips & I panicked & stopped it.
it was great fun but unmanageable in a live scenario.
I'd like to be able to set a hard limit on the number of tracks it could create like this, perhaps then forcing the user to plan the overdubs, & helping the user to avoid having to navigate through hundreds of similarly-named audio clips.

**Reply by kozmikyak** [reply]
That's a data point in favor of some of the people who've recommended that overdubs be destructive past the last overdub--in that you can undo one level of overdub, but not multiple.  Hardware loop pedals for this kind of thing offer that interface to "tracks" and while the Deluge's way is more flexible and powerful, grouping dubs into a "virtual track" of some kind or doing destructive overdubs can mitigate this somewhat.
